Propagate generic types where possible

joelpurra opened the pull request:
- Replaces the hardcoded `unknown` type in for example `is.promise(value)` with `is.promise<T>(value)` so that a known (inferred or explicit) generic type `T` can be propagated in the type system.
  - The same goes for `Key` and `Value` types in for example `Map<Key, Value>`, and so on.
  - Also replacing `{[key: string]: unknown}` and similar with
`Record<Key, Value>` as appropriate.
  - Kept `is.plainObject` as `Record<string, Value>`, although the
`Key` can also be a `symbol`.
- If no inferred/explicit generic type is provided, the (current) default/implicit generic type is preserved. It is usually `unknown`.
- Use with caution, as `is` (nor the associated assertions) can not check generic type validity at runtime. It is a compile-time check performed by typescript, based on the developer-provided types (inferred or explicit).
- Making the wrong compile-time assumption about `T` can result in value with types which differ at runtime, leading to unexpected behavior. It is safer to stick with `unknown` generic types until they have been verified at runtime. This is particularly important when it comes to verifying the types of arbitrary inputs from users and external systems.
- No tests added, as changes should be compile-time only.

See

- https://www.typescriptlang.org/docs/handbook/generics.html
- https://github.com/sindresorhus/is/issues/102

sindresorhus commented on source/index.ts:
```diff
@@ -590,12 +590,12 @@ export const assert: Assert = {
 	emptyStringOrWhitespace: (value: unknown): asserts value is string => assertType(is.emptyStringOrWhitespace(value), AssertionTypeDescription.emptyStringOrWhitespace, value),
-	emptyObject: (value: unknown): asserts value is {[key: string]: never} => assertType(is.emptyObject(value), AssertionTypeDescription.emptyObject, value),
-	nonEmptyObject: (value: unknown): asserts value is {[key: string]: unknown} => assertType(is.nonEmptyObject(value), AssertionTypeDescription.nonEmptyObject, value),
+	emptyObject: <Key extends keyof any = string>(value: unknown): asserts value is Record<Key, never> => assertType(is.emptyObject(value), AssertionTypeDescription.emptyObject, value),
```
Why `any` and not `unknown`?

joelpurra replied:
To match `Record<,>`. Also, `unknown` wouldn't work here as default types would not work. In this case it's reported as `Type 'string' does not satisfy the constraint 'never'.`

https://github.com/microsoft/TypeScript/blob/v3.7.5/lib/lib.es5.d.ts#L1466-L1471

sindresorhus replied:
👍

sindresorhus commented:
> Use with caution, as is (nor the associated assertions) can not check generic type validity at runtime. It is a compile-time check performed by typescript, based on the developer-provided types (inferred or implicit).

We should definitely have a warning in the readme about this.

brandon93s commented:
👍 from me.  Thank you for the examples in https://github.com/sindresorhus/is/issues/102.  Please add the above noted warning to the readme.

gioragutt commented:
@sindresorhus regarding the warning, it's true for anything that doesn't really from from your code, like `JSON.parse`ing a string... It's not really an `is` thing, but a typescript thing.

I do agree it's important, just in case.

joelpurra commented:
- Added a section about propagating generic type parameters in the readme.
- Added a link at the top of the readme, including a brief warning.
- Included an example based on wrongly assuming/correctly asserting a resolved `Promise<T>`. It might not be the best example; am open to suggestions.
